File: prob4.js
//Joe McCourt
//4/14/12

// Project Euler problem 4
// A palindromic number reads the same both ways. 
// The largest palindrome made from the product of 
// two 2-digit numbers is 9009 = 91*99.
// Find the largest palindrome made from 
// the product of two 3-digit numbers.
function Prob4(){
	this.isPalindrome = function(strNum){
		var length = strNum.length;
		if(length === 1){return true;}
		if(strNum[0] === strNum[length-1]){
			if(length === 2){return true;}
			return this.isPalindrome(strNum.substr(1,length-2));
		}
		return false;
	};

	this.eval = function(){
		var highestPal = 1;
		//Iterate over every combination of two 3-digit numbers
		for(var number1 = 1; number1 < 1000; number1++){
			for(var number2 = number1; number2 < 1000; number2++){
				//If this number is higher than highest found
				if(number1*number2 > highestPal){
					if(this.isPalindrome((number1*number2).toString())){
						//This is a palindrome and highest found so set
						highestPal = number1*number2;
					}
				}
			}
		}
		return highestPal;
	};
}

File: go/prob4.go
// A palindromic number reads the same both ways.
// The largest palindrome made from the product of
// two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

package main

import (
	"fmt"
)

//Does not handle negative numbers
func decomposeBase10(num int64) []int {

	numDigits := 1
	i := int64(10)

	for i <= num {
		numDigits++
		i *= 10
	}

	decomp := make([]int, numDigits)

	i = int64(numDigits - 1)
	for num >= 10 {
		lower := num / 10
		decomp[i] = int(num - lower*10)
		num = lower
		i--
	}

	decomp[0] = int(num)

	return decomp
}

func isPalindrome(num []int) bool {

	for i := 0; i < len(num)/2; i++ {
		if num[i] != num[len(num)-i-1] {
			return false
		}
	}

	return true
}

func prob4() {

	//brute force
	max := int64(0)
	for x1 := 100; x1 <= 1000; x1++ {
		for x2 := x1; x2 <= 1000; x2++ {
			num := int64(x1 * x2)

			if num > max {
				if isPalindrome(decomposeBase10(num)) {
					//fmt.Println(x1, x2, num)
					max = num
				}
			}
		}
	}

	fmt.Println(max)
}
